pkg/keymanagementprovider/azurekeyvault/provider.go

```diff
@@ -186,39 +207,80 @@ func (s *akvKMProvider) GetCertificates(ctx context.Context) (map[keymanagementp
+				if cert.ID == nil || cert.Attributes == nil || cert.Attributes.Created == nil {
+					logger.GetLogger(ctx, logOpt).Warnf("certificate %s, version %s, found invalid certificate object, id, attributes or created time must not be nil", keyVaultCert.Name, cert.ID.Version())
+					continue
 				}
-				certBundle := certResponse.CertificateBundle
-				keyVaultCert.Version = getObjectVersion(*certBundle.KID)
-				isEnabled := *certBundle.Attributes.Enabled
-				lastRefreshed := startTime.Format(time.RFC3339)
-				certProperty := getStatusProperty(keyVaultCert.Name, keyVaultCert.Version, lastRefreshed, isEnabled)
-				certsStatus = append(certsStatus, certProperty)
-				mapKey := keymanagementprovider.KMPMapKey{Name: keyVaultCert.Name, Version: keyVaultCert.Version, Enabled: isEnabled}
-				keymanagementprovider.DeleteCertificateFromMap(s.resource, mapKey)
-				continue
+				versionInfo := VersionInfo{
+					Version: cert.ID.Version(),
+					Created: *cert.Attributes.Created,
+				}
+				versionHistory = append(versionHistory, versionInfo)
 			}
-			return nil, nil, fmt.Errorf("failed to get secret objectName:%s, objectVersion:%s, error: %w", keyVaultCert.Name, keyVaultCert.Version, err)
 		}
 
-		secretBundle := secretResponse.SecretBundle
-		isEnabled := *secretBundle.Attributes.Enabled
+		// Pager results are not sorted by created time, so we sort them here
+		// in ascending order (oldest to newest)
+		sortVersionHistory(versionHistory)
+		sortedVersionHistory := GetVersions(versionHistory)
 
-		certResult, certProperty, err := getCertsFromSecretBundle(ctx, secretBundle, keyVaultCert.Name, isEnabled)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to get certificates from secret bundle:%w", err)
+		if len(sortedVersionHistory) == 0 {
+			logger.GetLogger(ctx, logOpt).Infof("no versions found for certificate %s", keyVaultCert.Name)
+			continue
 		}
 
-		metrics.ReportAKVCertificateDuration(ctx, time.Since(startTime).Milliseconds(), keyVaultCert.Name)
-		certsStatus = append(certsStatus, certProperty...)
-		certMapKey := keymanagementprovider.KMPMapKey{Name: keyVaultCert.Name, Version: keyVaultCert.Version, Enabled: isEnabled}
-		certsMap[certMapKey] = certResult
+		// if versionHistoryLimit isn't set default to 1 to avoid out of bounds error
+		if keyVaultCert.VersionHistoryLimit == 0 {
+			keyVaultCert.VersionHistoryLimit = versionHistoryLimitDefault
+		}
+
+		// Ensure that the versionHistoryLimit is not greater than the number of versions to avoid out of bounds error
+		if keyVaultCert.VersionHistoryLimit > len(sortedVersionHistory) {
+			keyVaultCert.VersionHistoryLimit = len(sortedVersionHistory)
+		}
+
+		// get the latest version of the certificate up to the limit
+		for _, version := range sortedVersionHistory[len(sortedVersionHistory)-keyVaultCert.VersionHistoryLimit:] {
+			secretReponse, err := s.secretKVClient.GetSecret(ctx, keyVaultCert.Name, version)
+			if err != nil {
+				if isSecretDisabledError(err) {
```

Comment:
Excellent call out @shizhMSFT! The `NewListCertificateVersionPager` does have the `Enabled` attribute, which can be used instead of the `isSecretDisabledError` func. However, I don't think we can fail much sooner. `n` certs means enabled and disabled certs. Meaning we won't know which disabled certs to write to the status until after the pager results are returned and sorted.

Reply:
I updated the fetchVersionHistory to use the enabled field from the pager to fail faster. However, the single fetch logic remains the same since it doesn't have access to the pager attributes.